File: apps/clrk-web/src/features/optimizer/data/remote-optimizer-repository.ts
import { apiClient, getApiErrorMessage } from '#/lib/api-client'
import type { OptimizationLevel, OptimizationResult, CutSuggestion } from '../types'

export type OptimizeRequest = {
  level: OptimizationLevel
  from: string
  to: string
}

type OptimizeResponse = {
  currency: string
  level: OptimizationLevel
  totalCurrentSpend: number
  totalSavings: number
  suggestions: CutSuggestion[]
}

export async function optimizeBudget({ level, from, to }: OptimizeRequest): Promise<OptimizationResult> {
  try {
    const { data } = await apiClient.post<OptimizeResponse>('/api/receipts/optimize', {
      level,
      from,
      to,
    })

    return {
      currency: data.currency,
      level: data.level,
      totalCurrentSpend: data.totalCurrentSpend,
      totalSavings: data.totalSavings,
      suggestions: data.suggestions,
    }
  } catch (error) {
    throw new Error(getApiErrorMessage(error, 'Unable to optimize budget.'))
  }
}

File: apps/clrk-mobile/lib/features/optimizer/data/remote_optimizer_repository.dart
import 'package:dio/dio.dart';

import '../domains/optimization_result.dart';

class RemoteOptimizerRepository {
  const RemoteOptimizerRepository(this._dio);

  final Dio _dio;

  Future<OptimizationResult> optimize({
    required String level,
    required String from,
    required String to,
  }) async {
    final response = await _dio.post<Map<String, dynamic>>(
      '/api/receipts/optimize',
      data: {'level': level, 'from': from, 'to': to},
    );

    return OptimizationResult.fromJson(
      response.data ?? const <String, dynamic>{},
    );
  }
}
